analysis/src/cratesio.rs

```diff
@@ -0,0 +1,93 @@
+//! This module abstracts the communication with crates.io for a given crate
+
+// TODO: A cheaper way to interact with crates.io can be working with their
+// experimental database dump that is updated daily, https://crates.io/data-access,
+// which will enable us to avoid making http requests and dealing with rate limits
+
+use crates_io_api::{SyncClient, Error};
+use anyhow::Result;
+use tabled::Tabled;
+
+#[derive(Tabled, Default)]
+pub struct CratesioReport<'a> {
+    pub name: &'a str,
+    pub is_hosted: bool,
+    pub downloads: u64,
+    pub dependents: u64,
+}
+
+pub struct CratesioAnalyzer;
+
+impl CratesioAnalyzer {
+    pub fn analyze_cratesio<'a>(&self, name: &'a str) -> Result<CratesioReport<'a>> {
+        let client = SyncClient::new(
+            //TODO: replace repo URL with an email address to receive traffic notification
+            "User-Agent: Whackadep (https://github.com/diem/whackadep)",
+            std::time::Duration::from_millis(1000),
+        )?;
+
+        let is_hosted = match self.is_hosted(&client, name) {
+            Ok(success) => success,
+            Err(error) => return Err(error.into()),
+        };
```

Comment:
Lets tweak the semantics of this function. calling `analyze_crateio` on a crate which isn't published on crates.io should return an error. instead the caller should check if its published by looking at [`PackageMetadata::sorce`](https://docs.rs/guppy/0.9.0/guppy/graph/struct.PackageMetadata.html#method.source) and then calling [`PackageSource::is_crates_io`](https://docs.rs/guppy/0.9.0/guppy/graph/enum.PackageSource.html#method.is_crates_io)

Reply:
Actually this function should probably take in a `PackageMetadata` instead of `name: &str` and that logic can live in this function.